zkvm/src/blockchain/state.rs

```diff
@@ -0,0 +1,89 @@
+use bulletproofs::BulletproofGens;
+use musig::VerificationKey;
+use std::collections::{HashSet, VecDeque};
+
+use super::block::{Block, BlockHeader, BlockID};
+use super::errors::BlockchainError;
+use crate::{Entry, TxLog, VMError, UTXO};
+
+#[derive(Clone)]
+pub struct BlockchainState {
+    pub initial: BlockHeader,
+    pub tip: BlockHeader,
+    pub utxos: HashSet<UTXO>,
+    pub nonces: VecDeque<([u8; 32], u64)>, // xxx need fast lookup by anchor
+    pub ref_ids: VecDeque<BlockID>,        // xxx need fast lookup by blockID
+
+    pub initial_id: BlockID,
+}
+
+impl BlockchainState {
+    pub fn make_initial(timestamp_ms: u64, refscount: u64) -> BlockchainState {
+        let initialHeader = BlockHeader::make_initial(timestamp_ms, refscount);
+        BlockchainState {
+            initial: initialHeader.clone(),
+            initial_id: initialHeader.id(),
+            tip: initialHeader,
+            utxos: HashSet::new(),
+            nonces: VecDeque::new(),
+            ref_ids: VecDeque::new(),
+        }
+    }
+
+    pub fn apply_block<F>(
+        &self,
+        b: &Block,
+        bp_gens: &BulletproofGens,
+        key_agg_fn: F,
+    ) -> Result<BlockchainState, BlockchainError>
+    where
+        F: Clone + Fn(&[VerificationKey]) -> Result<VerificationKey, VMError>,
+    {
+        let txlogs = b.validate(&self.tip, bp_gens, key_agg_fn)?;
+        let mut new_state = self.clone();
+
+        // Remove expired nonces.
+        while let Some(nonce_pair) = new_state.nonces.front() {
```

Comment:
It seems like this will remove nonces from the set until we encounter one with timestamp >= blockheader timestamp. Is that the desired behavior? Seems like we would want to iterate through all nonces and remove ones with expired timestamp, but I might be misunderstanding

Reply:
Great catch! This logic presumes that `new_state.nonces` is sorted by expiration time. Rather than fix anything, I added a comment for future implementers about that being a requirement.